gui/src/ruler.hh

```diff
@@ -0,0 +1,189 @@
+#pragma once
+
+#include <gtk/gtk.h>
+#include <boost/smart_ptr.hpp>
+
+/**
+ * This class draws a ruler to a GtkDrawingArea.
+ * It is intended as a replacement for the old GTK2 ruler widget and is written
+ * to mimic that widget's behavior as close as possible.
+ */
+class Ruler
```

Comment:
**Testing tip:**
This class has two responsibilities: Computing where stuff should go, and drawing stuff. Of those, you especially want to (unit) test the computing stuff. The drawing stuff is best tested manually.

However, unittesting the computation stuff is hard, because all of that is private in this class. Hence, I recommend introducing a RuleCalculator class (or something), that contains all the computation stuff (publicly), but none of the drawing stuff. Then you can unittest that class, and it can be a private member of Ruler, such that it still doesn't expose any internals you don't want the outside world to see.